hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -159,6 +161,177 @@ def trivial(self):
+        self.__blocks = {}
+        self.__count = 0
+        self.__sparse = True
+
+    def add_endpoint(self, n):
+        """Add n to the list of enpoints."""
+        assert isinstance(self.owner, ConjectureData)
+
+        # We store endpoints as the smallest size integer we can get
+        # away with. When we try to add an endpoint that is too large,
+        # we upgrade the array to the smallest word size needed to store
+        # the new result.
+        try:
+            self.__endpoints.append(n)
+        except OverflowError:
+            self.__endpoints = array_or_list(
+                BIT_LENGTH_TO_ARRAY_CODES[n.bit_length()], self.__endpoints
+            )
+            self.__endpoints.append(n)
+
+    def transfer_ownership(self, new_owner):
+        """Used to move ``Blocks`` over to a ``ConjectureResult`` object
+        when that is read to be used and we no longer want to keep the
+        whole ``ConjectureData`` around."""
+        assert isinstance(new_owner, ConjectureResult)
+        self.owner = new_owner
+
+    def start(self, i):
+        """Equivalent to self[i].start."""
+        if i == 0:
+            return 0
+        else:
+            return self.__endpoints[i - 1]
+
+    def end(self, i):
+        """Equivalent to self[i].end."""
+        return self.__endpoints[i]
+
+    def bounds(self, i):
+        """Equivalent to self[i].bounds."""
+        return (self.start(i), self.end(i))
+
+    def all_bounds(self):
+        """Equivalent to [(b.start, b.end) for b in self]."""
+        result = []
+        prev = 0
+        for e in self.__endpoints:
+            result.append((prev, e))
+            prev = e
+        return result
+
+    def __len__(self):
+        return len(self.__endpoints)
+
+    def __known_block(self, i):
+        try:
+            return self.__blocks[i]
+        except (KeyError, IndexError):
+            return None
+
+    def __getitem__(self, i):
+        n = len(self)
+        if i < -n or i >= n:
+            raise IndexError("Index %d out of range [-%d, %d)" % (i, n, n))
+        if i < 0:
+            i += n
+        assert i >= 0
+        result = self.__known_block(i)
+        if result is not None:
+            return result
+
+        # We store the blocks as a sparse dict mapping indices to the
+        # actual result, but this isn't the best representation once we
+        # stop being sparse and want to use most of the blocks. Switch
+        # over to a list at that point.
+        if self.__sparse and len(self.__blocks) * 2 >= len(self.__endpoints):
+            new_blocks = [None] * len(self.__endpoints)
+            for k, v in self.__blocks.items():
+                new_blocks[k] = v
+            self.__sparse = False
+            self.__blocks = new_blocks
+            assert self.__blocks[i] is None
+
+        start = self.start(i)
+        end = self.__endpoints[i]
+        self.__count += 1
+        assert self.__count <= len(self.__endpoints)
```

Comment:
I have no idea why this should hold.

Reply:
`len(self.__endpoints)` is the total number of blocks that are logically in the block list, while `self.__count` keeps track of the number of `Block` instances that we have actually created so far.

(At this point in the code we've just incremented the count because we decided to create a new `Block`, but haven't actually created it yet.)

Every `Block` instance represents one of those logical blocks, so if there are more instances than actual blocks then something has gone out of sync between the two.